### DISCUSSION TITLE
What makes lima a better alternative to running a Linux VM on apps like Parallels or VMware Fusion? Are there any particular performance optimizations done for Lima?

### DISCUSSION THREAD

**Discussion by YousufSSyed** [lima-vm/lima#557] [Q&A] [answered]
Optimizations besides using QEMU with HVF. In terms of advantages, besides integration with macOS like port forwarding and access to the Mac home directory. Either way, I like how it's basically a Mac equivalent of Windows Subsystem for Linux, and will use it.

**Comment by afbjorklund** [comment, marked as answer]
Primarly I see Lima as a tool for setting up an environment to run containers on, like `docker-machine` or `podman-machine`

But Lima is more like [Vagrant](https://www.vagrantup.com/), it allows you to customize the VM a bit (if you want to). It also runs on Linux (KVM), great for developers!

I think [UTM](https://mac.getutm.app/) would be more like those other tools ? When it comes to "just" offering a GUI for QEMU (wonder what happened to QtEmu ?)

"Are there any particular performance optimizations done for Lima?" You can look at [Alpine Lima](https://github.com/lima-vm/alpine-lima), a custom Linux distribution.

**Reply by reorx** [reply]
Hi, I'm new to Lima, and the first question come to my mind is "what makes Lima different from Vagrant?", so I searched and found this discussion, can you explain this a bit?

**Reply by afbjorklund** [reply]
You can accomplish the same thing, the syntax is a bit different (YAML, instead of Ruby). In my configurations, most steps are handled by inline shell scripts or external programs anyway. So it's not a big deal, having to "translate" the rest ?

With Vagrant, I have normally been using the VirtualBox provider (that doesn't work on ARM) or the Libvirt/KVM provider (that doesn't work on Mac). While Lima uses the regular upstream Cloud images, instead of special "boxes".

There is a new version of Vagrant in progress, once it is completed it could be interesting to compare it with Lima:

https://www.hashicorp.com/blog/toward-vagrant-3-0

But Lima is available already today, it is Open Source and driven by an active community instead of a corporation.

What use cases and vagrant setups, did **you** have ?

**Reply by afbjorklund** [reply]
By the way, here are the vagrant files, that I was talking about above:
https://boot2podman.github.io/2020/07/22/machine-replacement.html

This was similar to "docker.yaml" and "podman.yaml" examples in Lima.
In a way, `limactl` is similar to a `nerdctl-machine` (for containerd/buildkit).

* Docker Machine: https://asciinema.org/a/366096

* Podman Machine: https://asciinema.org/a/366097

The main difference is that nerdctl doesn't have a remote API, only a local CLI.
So we run the container commands using SSH instead, such as `lima nerdctl`.

This means that all files need to be in a mounted directory, to be accessible...
Because the nerdctl client is actually running on the VM, instead of on the host.

**Reply by AkihiroSuda** [reply]
The biggest difference from Vagrant is automatic port forwarding.
The second biggest one is filesystem sharing, the third one is nerdctl integration.

It should be also noted that Vagrant doesn't work on ARM when it is configured to use VirtualBox driver.

**Comment by afbjorklund** [comment]
"What makes ~~lima~~ QEMU a better alternative to running a Linux VM on apps like Parallels or VMware Fusion?"

A very important part would be Open Source. There is also a lower cost, but they are not _that_ steep ($100 - $200)

For VirtualBox, I think the main answer is: ARM.

Otherwise, that was the _other_ go-to VM solution.

**Comment by andrewcrook** [comment]
When I go through the lima configuration templates for default it tells me that that writing is “dangerous” and only read mounts are recommended for home directory (‘/User’) . I would like a safe way for my trusted dev vms to read/right to folders in my Mac home directory. Parallels seems to be very stable in this regard and offers no such warning.  I wonder how difficult would it be to add parallels support via a vmtype to lima?

```
# 🟢 Builtin default: null (Mount nothing)
# 🔵 This file: Mount the home as read-only, /tmp/lima as writable
mounts:
- location: "~"
  # Configure the mountPoint inside the guest.
  # 🟢 Builtin default: value of location
  mountPoint: null
  # CAUTION: `writable` SHOULD be false for the home directory.
  # Setting `writable` to true is possible, but untested and dangerous.
  # 🟢 Builtin default: false
  writable: null
```

```
 Mount type for above mounts, such as "reverse-sshfs" (from sshocker), "9p" (EXPERIMENTAL, from QEMU’s virtio-9p-pci, aka virtfs),
# or "virtiofs" (EXPERIMENTAL, needs `vmType: vz`)
# 🟢 Builtin default: "reverse-sshfs"
mountType: null
```

**Reply by AkihiroSuda** [reply]
Probably we will remove `dangerous` warning in Lima v1.0 soon.
